consensus/tendermint/router/routing.go

```diff
@@ -0,0 +1,377 @@
+			m.latestLatencies[addr] = lat
+			updated = true
+		}
+	}
+	m.latencyMu.Unlock()
+
+	m.retryMu.Lock()
+	m.nodesToRetry = make(map[common.Address]struct{})
+	for _, addr := range failedNodes {
+		m.nodesToRetry[addr] = struct{}{}
+	}
+	m.retryMu.Unlock()
+
+	if updated {
+		m.latencyMu.Lock()
+		latMap := make(map[common.Address]uint, len(m.latestLatencies))
+		for addr, lat := range m.latestLatencies {
+			latMap[addr] = lat
+		}
+		m.latencyMu.Unlock()
+		m.refreshClustersLatencies(latMap)
+		log.Debug("Router: clusters updated after latency retry")
+	}
+
+	log.Debug("Router: latency retry completed", "failed_nodes", len(failedNodes))
+	return nil
+}
+
+func (m *Router) loop(ctx context.Context) {
+	defer m.wg.Done()
+
+	retryTicker := time.NewTicker(constants.RetryLatencyTimeout)
+	cleanupTicker := time.NewTicker(constants.CacheCleanupInterval)
+	// wait for few seconds before starting the initial latency measurement
+	initialMeasurementTimer := time.NewTimer(5 * time.Second)
+	defer func() {
+		retryTicker.Stop()
+		cleanupTicker.Stop()
+		initialMeasurementTimer.Stop()
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-initialMeasurementTimer.C:
+			if !m.inCommittee || m.peerFinder == nil || len(m.committee) < m.clusteringThreshold {
+				continue
+			}
+			log.Debug("Router: initial latency measurement started", "threshold", m.clusteringThreshold)
+			if err := m.measureLatency(); err != nil {
+				log.Warn("measureToReport failed", "err", err)
+			}
+			initialMeasurementTimer.C = nil
+		case <-time.After(constants.LatencyDataExpiry +
+			time.Duration(rand.Intn(constants.LatencyMeasurementDelayCap))*time.Millisecond):
+			if !m.inCommittee || m.peerFinder == nil || len(m.committee) < m.clusteringThreshold {
+				continue
+			}
+			if err := m.measureLatency(); err != nil {
+				log.Warn("measureToReport failed", "err", err)
+			}
+		case <-retryTicker.C:
+			retryTicker = time.NewTicker(constants.RetryLatencyTimeout)
+			if !m.inCommittee || m.peerFinder == nil || len(m.committee) < m.clusteringThreshold {
+				continue
+			}
+			if err := m.retryLatency(); err != nil {
+				log.Warn("retryLatency failed", "err", err)
+			}
+		case <-cleanupTicker.C:
+			m.recipientCache.Cleanup()
+			log.Debug("Router: cache cleanup completed")
+		case epochEv := <-m.epochEventChan:
+			log.Info("Router: new epoch detected", "height", epochEv.Header.Number.String())
+			epoch := epochEv.Header.Epoch
+			m.inCommittee = epoch.Committee.MemberByAddress(m.self) != nil
+			if !m.inCommittee || len(m.committee) < m.clusteringThreshold {
+				log.Info("Router: clustering not needed, skipping measurement")
+				continue
+			}
+			m.updateCommittee(epoch)
+			nw, err := network.New(m.committee, m.Latencies(), m.self)
+			if err != nil {
+				log.Error("Router: failed to create network", "err", err)
+				continue
+			}
+			m.updateNetwork(nw)
```

Comment:
I think we want the following instead. Because even if we get error, we might still want to call `m.measureLatency()` in the following line.

```suggestion
			m.refreshClustersLatencies(m.Latencies())
```